### PR TITLE
Add bnb solver and clean solvers

### LINKED ISSUE #28: Add BnB solver

### PR DESCRIPTION
**Pull request by artturuusila**
closes #28

### CODE REVIEW

File: soft4pes/control/mpc/solvers/mpc_bnb.py [hunk 1]
```diff
@@ -0,0 +1,129 @@
...
+
+                    # If not at the last prediction step, move to the next prediction step
+                    if k < ctr.Np - 1:
+                        self.u_temp[3 * k:3 * (k + 1)] = uk
+                        self.solve(sys, conv, ctr, x_kp1, y_ref, uk, k + 1,
+                                   J_temp)
+                    else:
+                        # If at the last prediction step, store the three-phase switch position and
+                        # update the minimum cost
+                        self.U_temp[3 * k:3 * (k + 1)] = uk
```
Comment (PetPKar) [inline soft4pes/control/mpc/solvers/mpc_bnb.py:127]: Is this u(k) or u(k+Np-1)?

File: soft4pes/control/mpc/solvers/mpc_enum.py [hunk 2]
```diff
@@ -102,10 +103,10 @@ def solve(self, sys, conv, ctr, xk, y_ref, u_km1, k):
...
 
-        # Iterate over all possible 3-phase switch position
+        # Iterate over all possible three-phase switch positions
         for uk in self.SW_COMB:
             # Check if switching constraint is violated or cost is infinite
-            if self.switching_constraint_violated(
+            if switching_constraint_violated(
                     conv.nl, uk, u_km1) or self.J[self.i] == np.inf:
                 # Set the cost to infinity and use the current state for the next
                 # prediction step
```
[the remark this answers is not in the document]
Comment (artturuusila) [reply soft4pes/control/mpc/solvers/mpc_enum.py:112]: Corrected in the next commit

File: soft4pes/control/mpc/solvers/mpc_enum.py [hunk 3]
```diff
@@ -116,50 +117,23 @@ def solve(self, sys, conv, ctr, xk, y_ref, u_km1, k):
...
                     (len(u_row_range), 1)) * uk
                 self.J[u_row_range] = self.J[self.i]
 
                 # Move to the next prediction step
                 self.solve(sys, conv, ctr, x_kp1, y_ref, uk, k + 1)
             else:
-                # If at the last prediction step, store the 3-phase switch position
-                self.u_seq[self.i, u_col_range_start:u_col_range_end] = uk
+                # If at the last prediction step, store the three-phase switch position
+                self.U_seq[self.i, u_col_range_start:u_col_range_end] = uk
```
[the remark this answers is not in the document]
Comment (artturuusila) [reply soft4pes/control/mpc/solvers/mpc_enum.py:138]: Yes, this line selects U_seq( 3*k : 3(k+1) ). So for example, if Np = 3, U_seq will have 9 entries. In the last step, k = 2 (0-based indexing). We will select range [6 : 9), and assign the last three-phase switch uk position there.

File: soft4pes/control/mpc/solvers/mpc_bnb.py [hunk 4]
```diff
@@ -7,18 +7,18 @@
...
     Attributes
     ----------
     J_min : float
         Minimum cost.
-    u_seq : 1 x 3*Np ndarray
-        Array for the 3-phase switch position sequence with the lowest cost.
-    u_temp : 1 x 3*Np ndarray
-        Temporary array for 3-phase switch position sequence.
+    U_seq : 1 x 3*Np ndarray
+        Array for sequences of three-phase switch positions (switching sequences) with the lowest cost.
```
Comment (artturuusila) [inline soft4pes/control/mpc/solvers/mpc_bnb.py:17]: I have to ask, you modified this to be sequences. Isn't it that we have one sequence of three-phase switch positions? In the enumeration, we store multiple sequences. Is this a mistake or do I understand something wrong?

Comment (PetPKar) [reply soft4pes/control/mpc/solvers/mpc_bnb.py:17]: u(k) is a three-phase switch position
if we have {u(k), u(k+1), u(k+2), ...} we have a switching sequence

File: soft4pes/control/mpc/solvers/mpc_enum.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (PetPKar) [reply soft4pes/control/mpc/solvers/mpc_enum.py:138]: But then do you assign the last three-phase position of the sequence (that u(k+2) in your example) to be u(k), i.e., the applied switch position? In other words, which is the applied switch position? The first three entries of the sequence or the last?

File: soft4pes/control/mpc/solvers/mpc_enum.py [hunk 5]
```diff
@@ -116,50 +117,23 @@ def solve(self, sys, conv, ctr, xk, y_ref, u_km1, k):
...
+                y_error = np.linalg.norm(y_ref[k] - np.dot(ctr.C, x_kp1))**2
                 delta_u = ctr.lambda_u * np.linalg.norm(uk - u_km1, ord=1)
-                self.J[self.i] = self.J[self.i] + ref_error + delta_u
+                self.J[self.i] = self.J[self.i] + y_error + delta_u
 
             if k < ctr.Np - 1:
                 # If not at the last prediction step, recursively call mpc_enum
-                # Assign the used 3-phase switch position to all reduntant switch position sequences
-                # Otherwise, try the next 3-phase switch position
+                # Assign the used three-phase switch position to all reduntant switching sequences
```
[the remark this answers is not in the document]
Comment (PetPKar) [reply soft4pes/control/mpc/solvers/mpc_enum.py]: The comment was mostly about the choice of the word "redundant". Are these redundant sequences? In what sense? I think this needs to be explained

File: soft4pes/control/mpc/solvers/mpc_bnb.py (hunk 1 above)
Comment (PetPKar) [reply soft4pes/control/mpc/solvers/mpc_bnb.py:127]: I think this relates to the comment I made above about which part of the sequence is considered as the applied three-phase switch position.

As for the naming, I think this needs to be fixed as soon as possible. At the moment, some definitions look confusing (or contradictory) and it seems that some comments will just re-appear as long as this issue has not been addressed

**Comment by PetPKar** [comment]
Some clarifications in notation/terminology are needed to avoid confusion and ensure understanding in the future

File: soft4pes/control/mpc/solvers/mpc_bnb.py [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+            System model.
+        conv : object
+            Converter model.
+        ctr : object
+            Controller object.
+        x_ell : ndarray of floats
+            State vector [p.u.].
+        y_ref : ndarray of floats
+            Reference vector [p.u.].
+        u_ell_prev : 1 x 3 ndarray of ints
```
Comment (PetPKar) [inline soft4pes/control/mpc/solvers/mpc_bnb.py:92]: for description "prev" and "next", see comment in the ``mpc_enum`` file

**Review by PetPKar — COMMENTED** [review]
Things look good. Just a few small comments on the notation

File: soft4pes/control/mpc/controllers/rl_grid_mpc_curr_ctr.py [hunk 7]
```diff
@@ -85,7 +85,7 @@ def __call__(self, sys, conv, t):
         Returns
```
[the remark this answers is not in the document]
Comment (artturuusila) [reply soft4pes/control/mpc/controllers/rl_grid_mpc_curr_ctr.py:85]: Yes, I have a note about using kTs instead of t. This is a matter of different PR

**Comment by PetPKar** [comment]
Things look ok

**Review by PetPKar — APPROVED** [review]

**Pull request merged**